### PR TITLE
Exclude added flash types from action_methods

### LINKED ISSUE #33593: Added flash types are listed in controller action methods
### Steps to reproduce

1. Subclass `ActionController::Base`
2. Add an additional flash type via `add_flash_types`
3. Inspect that controller's `action_methods`
4. You'll see your flash type included in that set

``` ruby
# frozen_string_literal: true

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails"
end

require "action_controller/railtie"

class TestApp < Rails::Application
  config.root = __dir__
  secrets.secret_key_base = "secret_key_base"

  config.logger = Logger.new($stdout)
  Rails.logger  = config.logger

  routes.draw do
    get "/" => "test#index"
  end
end

class TestController < ActionController::Base
  add_flash_types :hype
end

require "minitest/autorun"
require "rack/test"

class BugTest < Minitest::Test
  include Rack::Test::Methods

  def test_returns_success
    assert TestController.action_methods.include?('hype')
  end
end

```

### Expected behavior

Flash types should not be listed as action methods.

### Actual behavior
Flash types are indeed being listed as action methods.

### System configuration
**Rails version**:
Observed in 5.2.1 and edge.

**Ruby version**:
2.5.1

### PR DESCRIPTION
**Pull request by Bodacious**
Exclude additional flash types from action_methods.

Fixes #33593
Copy of stale PR #33594

When new flash types are added to ApplicationController, these
should not appear in the action_methods set, since they're
not actions. This commit corrects this behaviour.

``` ruby
class MyController < ApplicationController
  add_flash_types :hype
end

# Currently:
MyController.action_methods.include?('hype') # => true

# This commit
MyController.action_methods.include?('hype') # => false
```

### CODE REVIEW

File: actionpack/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,16 @@
+*   Exclude additional flash types from `ActionController::Base.action_methods`
```
Comment (eugeneius) [inline actionpack/CHANGELOG.md:1]:
```suggestion
*   Exclude additional flash types from `ActionController::Base.action_methods`.
```

File: actionpack/lib/abstract_controller/base.rb [hunk 2]
```diff
@@ -88,14 +88,13 @@ def internal_methods
       # * <tt>Set</tt> - A set of all methods that should be considered actions.
       def action_methods
         @action_methods ||= begin
+
```
Comment (eugeneius) [inline actionpack/lib/abstract_controller/base.rb:91]:
```suggestion
```

File: actionpack/lib/abstract_controller/base.rb [hunk 3]
```diff
@@ -88,14 +88,13 @@ def internal_methods
...
           # All public instance methods of this class, including ancestors
           methods = (public_instance_methods(true) -
             # Except for public instance methods of Base and its ancestors
             internal_methods +
             # Be sure to include shadowed public instance methods of this class
-            public_instance_methods(false))
-
-          methods.map!(&:to_s)
+            public_instance_methods(false)).map(&:to_s)
```
Comment (eugeneius) [inline actionpack/lib/abstract_controller/base.rb:97]: This regresses a performance optimisation from https://github.com/rails/rails/pull/33693.

```suggestion
            public_instance_methods(false))
            
          methods.map!(&:to_s)
```

File: actionpack/lib/action_controller/metal/flash.rb [hunk 4]
```diff
@@ -37,10 +37,15 @@ def add_flash_types(*types)
             request.flash[type]
           end
           helper_method(type) if respond_to?(:helper_method)
```
Comment (eugeneius) [inline actionpack/lib/action_controller/metal/flash.rb:39]:
```suggestion
          helper_method(type) if respond_to?(:helper_method)

```

File: actionpack/lib/action_controller/metal/flash.rb [hunk 5]
```diff
@@ -37,10 +37,15 @@ def add_flash_types(*types)
...
-
           self._flash_types += [type]
         end
       end
+
+      # Extend AbstractController::Base::action_methods to remove any added flash types
+      # from the set
```
Comment (eugeneius) [inline actionpack/lib/action_controller/metal/flash.rb:45]: I think this method is simple enough that we don't need internal documentation for it.

```suggestion
```

File: actionpack/lib/action_controller/metal/flash.rb [hunk 6]
```diff
@@ -37,10 +37,15 @@ def add_flash_types(*types)
...
+      def action_methods #:nodoc:
+        super - _flash_types.map(&:to_s).to_set
+      end
```
Comment (eugeneius) [inline actionpack/lib/action_controller/metal/flash.rb:48]: It's important that `action_methods` is memoised, since it's called during every request.

We can copy the approach from the `UrlFor` module and use the same instance variable name, so that `clear_action_methods!` still works:

https://github.com/rails/rails/blob/a2cb0b23b7233371d8a2e0f843885dae330da2b7/actionpack/lib/abstract_controller/url_for.rb#L24-L30

```suggestion
      def action_methods #:nodoc:
        @action_methods ||= super - _flash_types.map(&:to_s).to_set
      end
```

**Review by eugeneius — CHANGES_REQUESTED** [review]

**Comment by eugeneius** [comment]
Can you squash your commits too?

File: actionpack/lib/abstract_controller/base.rb (hunk 3 above)
Comment (Bodacious) [reply actionpack/lib/abstract_controller/base.rb]: Well caught, thank you

**Comment by Bodacious** [comment]
@eugeneius I've applied the suggested changes. 

I rebased my PR against main and squashed the commits. Sorry, I wasn't sure if this project squashes on merge or not.

**Comment by p8** [comment]
@Bodacious Could you please rebase on main again?
The spellcheck should be fixed, making this build green again.

File: actionpack/lib/abstract_controller/base.rb [hunk 7]
```diff
@@ -96,7 +97,6 @@ def action_methods
             public_instance_methods(false))
 
           methods.map!(&:to_s)
-
```
Comment (eugeneius) [inline actionpack/lib/abstract_controller/base.rb:99]: There are still unnecessary changes in this file - we try to avoid those to keep the git history clean.

Can you also clean up the commit message, it contains all of the squashed commits' messages (e.g. "Add suggested changes from feedback").

**Comment by Bodacious** [comment]
@eugeneius Thanks. I've removed that extra whitespace and cleaned up the commit message

File: actionpack/lib/abstract_controller/base.rb [hunk 8]
```diff
@@ -96,7 +96,6 @@ def action_methods
...
```
Comment (eugeneius) [inline actionpack/lib/abstract_controller/base.rb:99]: Can you restore this line break? Otherwise I think this is good to go.

Comment (Bodacious) [reply actionpack/lib/abstract_controller/base.rb]: Sorry, I missed that

**Comment by eugeneius** [comment]
Thank you @Bodacious!

**Comment by p8** [comment]
Thanks @Bodacious!

Regarding this comment:
> > Can't we fix this problem by defining the flash methods as private?

> That would be a much nearer solution, but would break existing code that calls those methods on the controller.

https://github.com/rails/rails/pull/33594#issuecomment-425505321

I'm not sure where anyone would call those methods on the controllers...

**Comment by Bodacious** [comment]
@p8 It's bad practice, but it's not difficult to imaging some existing applications calling something like `controller.notice` from helpers or views directly, or `self.notice` within a controller.

Making the flash private would be a breaking change for these cases.

**Comment by p8** [comment]
@Bodacious Ah thanks, I hadn't thought about calling `controller.notice` 👍

**Comment by 2called-chaos** [comment]
To be fair I wasn't aware of controller.notice/alert but is it therefore fact that these are reserved action names? Prior to Rails 7 you could have actions named this way and work this way

    get '/test' => 'test#alert'

This stops working with Rails 7 but in a non-intuitive way, it just claims to not find that action despite it being listed in /rails/info/routes and you really have to dig to figure out why.

[via reddit](https://www.reddit.com/r/rails/comments/s8tiud/rails_7_are_notice_and_alert_actions_now_reserved/)

**Comment by ghiculescu** [comment]
FYI this has had an unexpected side effect - see https://github.com/rails/rails/issues/44867
